ARC: codegen bug, generated cpp doesn't compile

Test casem compile with `nim cpp --gc:arc XXX`
```nim
type
  TokenKind* = enum
    tkColon
    tkComma
    tkString
    tkNumber
    tkInt64
    tkIdent

  Token* = object
    case kind*: TokenKind
    of tkString: strVal*: string
    of tkNumber: numVal*: float
    of tkInt64: int64Val*: int64
    of tkIdent: ident*: string
    else: discard
    pos*: Natural

  BaseLexer* = object of RootObj
    input*: string
    pos*: Natural

  Json5Lexer* = object of BaseLexer

  JsonLexer* = object of BaseLexer
    allowComments*: bool
    allowSpecialFloats*: bool

  Lexer* = Json5Lexer | JsonLexer

  Parser[T: Lexer] = object
    l: T
    tok: Token
    allowTrailingComma: bool
    allowIdentifierObjectKey: bool

proc initJson5Lexer*(input: string): Json5Lexer =
  result.input = input

proc parseJson5*(input: string)  =
  var p = Parser[Json5Lexer](
    l: initJson5Lexer(input),
    allowTrailingComma: true,
    allowIdentifierObjectKey: true
  )

let x = "string"
parseJson5(x)
```

Generated cpp code fails compilation
```
Error: execution of an external compiler program 'g++.exe -c  -w -mno-ms-bitfields -w -fpermissive -mno-ms-bitfields   -IC:\Nim\lib -IC:\Nim -o "C:\nimcache\x1_d\@mx1.nim.cpp.o" "C:\nimcache\x1_d\@mx1.nim.cpp"' failed with exit code: 1

C:\nimcache\x1_d\@mx1.nim.cpp:114:147: error: could not convert '{{(& NTI__Tnwtohn42DJrL2jMvY4yjg_)}}' from '<brace-enclosed initializer list>' to 'tyObject_Json5Lexer__Tnwtohn42DJrL2jMvY4yjg'
 NIM_CONST tyObject_Parser__NcVQJhpnW9czydG9crD2Zbsg TM__ZPpeHLXgWrJ9bLTF8Y8dgsg_4 = {{{(&NTI__Tnwtohn42DJrL2jMvY4yjg_)}}, {}, NIM_FALSE, NIM_FALSE}
```

Example is from https://github.com/GULPF/samson library

--- a/lib/pure/smtp.nim
+++ b/lib/pure/smtp.nim
@@ -60,6 +60,7 @@ type
 
   SmtpBase[SocketType] = ref object
     sock: SocketType
+    address: string
     debug: bool
 
   Smtp* = SmtpBase[Socket]
@@ -225,15 +226,19 @@ proc checkReply*(smtp: Smtp | AsyncSmtp, reply: string) {.multisync.} =
   if not line.startswith(reply):
     await quitExcpt(smtp, "Expected " & reply & " reply, got: " & line)
 
+proc helo*(smtp: Smtp | AsyncSmtp) {.multisync.} =
+  # Sends the HELO request
+  await smtp.debugSend("HELO " & smtp.address & "\c\L")
+  await smtp.checkReply("250")
+
 proc connect*(smtp: Smtp | AsyncSmtp,
               address: string, port: Port) {.multisync.} =
   ## Establishes a connection with a SMTP server.
   ## May fail with ReplyError or with a socket error.
+  smtp.address = address
   await smtp.sock.connect(address, port)
-
   await smtp.checkReply("220")
-  await smtp.debugSend("HELO " & address & "\c\L")
-  await smtp.checkReply("250")
+  await smtp.helo()
 
 proc startTls*(smtp: Smtp | AsyncSmtp, sslContext: SSLContext = nil) {.multisync.} =
   ## Put the SMTP connection in TLS (Transport Layer Security) mode.
@@ -245,6 +250,7 @@ proc startTls*(smtp: Smtp | AsyncSmtp, sslContext: SSLContext = nil) {.multisync
       getSSLContext().wrapConnectedSocket(smtp.sock, handshakeAsClient)
     else:
       sslContext.wrapConnectedSocket(smtp.sock, handshakeAsClient)
+    await smtp.helo()
   else:
     {.error: "SMTP module compiled without SSL support".}
 

